pkg/apis/migration/migration.go

```diff
@@ -395,12 +428,30 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
 		createSpec.Profile = append(createSpec.Profile, profileSpec)
 	}
 	for _, datacenter := range datacenterPaths {
+		// Check vCenter API Version
 		// Format:
 		// https://<vc_ip>/folder/<vm_vmdk_path>?dcPath=<datacenter-path>&dsName=<datastoreName>
 		backingDiskURLPath := "https://" + host + "/folder/" +
 			vmdkPath + "?dcPath=" + url.PathEscape(datacenter) + "&dsName=" + url.PathEscape(datastoreName)
-		createSpec.BackingObjectDetails = &cnstypes.CnsBlockBackingDetails{BackingDiskUrlPath: backingDiskURLPath}
-		log.Infof("Registering volume: %q using backingDiskURLPath :%q", volumeSpec.VolumePath, backingDiskURLPath)
+		bUseVslmAPIs, err := common.UseVslmAPIs(ctx, vCenter.Client.ServiceContent.About.ApiVersion)
+		if err != nil {
+			msg := fmt.Sprintf("Error while determining the correct APIs to use for vSphere version %q, Error= %+v", vCenter.Client.ServiceContent.About.ApiVersion, err)
+			log.Errorf(msg)
+			return "", errors.New(msg)
+		}
+		if bUseVslmAPIs {
+			backingObjectID, err := (*volumeMigration.volumeManager).RegisterDisk(ctx, backingDiskURLPath, volumeSpec.VolumePath)
+			if err != nil {
+				msg := fmt.Sprintf("registration failed for volumePath: %v", volumeSpec.VolumePath)
+				log.Error(msg)
+				return "", errors.New(msg)
+			}
+			createSpec.BackingObjectDetails = &cnstypes.CnsBlockBackingDetails{BackingDiskId: backingObjectID}
```

Comment:
Did you observe any issue while tagging volume as CNS, right after registering vmdk as FCD? I remember we had some sync internal requirement.

Reply:
I tested on both 67u3 and 7.0 and I did not observe sync issues. I can test it again with more volumes. Let me get back on this

Reply:
I verified with 10 VCP PVCs with different sizes and observed no sync issues.
```
root@k8s-master:~# kubectl get pvc -A
NAMESPACE   NAME       STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
default     vcppvc     Bound    pvc-75a78f87-0816-4e2a-b02d-0a91ceeb76a4   10Mi       RWO            vcpsc          36m
default     vcppvc1    Bound    pvc-f35aa7b1-adef-4701-9123-7b2d720a7597   15Mi       RWO            vcpsc          27m
default     vcppvc10   Bound    pvc-e3738adb-c569-4ed0-9773-27d2a2760f42   10Gi       RWO            vcpsc          19m
default     vcppvc2    Bound    pvc-3aede78b-f1e8-45e6-86d0-f070287c75d2   15Mi       RWO            vcpsc          21m
default     vcppvc3    Bound    pvc-f2bbccd9-898c-4338-94dc-913f5d2e4a37   15Mi       RWO            vcpsc          21m
default     vcppvc4    Bound    pvc-be57b788-f24a-438e-9f39-d3e0a8db6c86   15Mi       RWO            vcpsc          21m
default     vcppvc5    Bound    pvc-6be41ebb-3b10-4207-8af2-be8bf9a2448f   15Mi       RWO            vcpsc          20m
default     vcppvc6    Bound    pvc-bbf4b95c-9966-445f-b5aa-41636c2d2967   150Mi      RWO            vcpsc          20m
default     vcppvc7    Bound    pvc-cd9f926c-a8b3-43a9-8221-98f4103640dc   1Gi        RWO            vcpsc          20m
default     vcppvc8    Bound    pvc-5d9c7764-70c6-4bd6-82da-6148ca944708   5Gi        RWO            vcpsc          20m
default     vcppvc9    Bound    pvc-47ae19a5-3597-4c84-ad25-4bb02fdc8c3d   8Gi        RWO            vcpsc          19m
```

Enabled CSI migration:
```
kubectl get cnsvspherevolumemigration -A
NAME                                   AGE
23d25b1f-ddf9-455b-ac5b-194265b19cab   7m58s
28fbb7dc-67c6-47a6-955d-38b2e806419f   6m50s
4034afce-322a-4a05-ad57-411cefbd98aa   7m27s
43289c14-c157-49e8-a9b8-13508176e730   7m8s
51851516-3c53-42d7-b25c-dd7f8848107f   6m44s
5de132d8-9cad-41b0-b87b-d41e47094022   7m1s
90874f52-ed66-4bd4-8ccd-12f9ba61d7e3   6m55s
9e054761-4666-45a5-b47c-bea59a49a1f2   7m34s
aea7244e-9a12-4616-99f1-7f3b29d2d3f9   7m20s
d3797a25-7dee-46e8-99ef-2e7f3cce5194   7m23s
ead503b7-5915-4a3e-bcd8-b06d3b13ad25   7m15s
```

Logs:
```
kubectl logs vsphere-csi-controller-68ccf5ddf4-75pwj -c vsphere-syncer -n vmware-system-csi | grep "using backingDiskId"
{"level":"info","time":"2021-04-01T18:29:03.041675223Z","caller":"migration/migration.go:450","msg":"Registering volume: \"[vsanDatastore] 4bb55260-de8a-c071-913b-020049bb8da4/kubernetes-dynamic-pvc-3aede78b-f1e8-45e6-86d0-f070287c75d2.vmdk\" using backingDiskId :\"23d25b1f-ddf9-455b-ac5b-194265b19cab\""}
{"level":"info","time":"2021-04-01T18:29:09.594198129Z","caller":"migration/migration.go:450","msg":"Registering volume: \"[vsanDatastore] 4bb55260-de8a-c071-913b-020049bb8da4/kubernetes-dynamic-pvc-47ae19a5-3597-4c84-ad25-4bb02fdc8c3d.vmdk\" using backingDiskId :\"9e054761-4666-45a5-b47c-bea59a49a1f2\""}
{"level":"info","time":"2021-04-01T18:29:15.822341641Z","caller":"migration/migration.go:450","msg":"Registering volume: \"[vsanDatastore] 4bb55260-de8a-c071-913b-020049bb8da4/kubernetes-dynamic-pvc-6be41ebb-3b10-4207-8af2-be8bf9a2448f.vmdk\" using backingDiskId :\"4034afce-322a-4a05-ad57-411cefbd98aa\""}
{"level":"info","time":"2021-04-01T18:29:21.984600357Z","caller":"migration/migration.go:450","msg":"Registering volume: \"[vsanDatastore] 4bb55260-de8a-c071-913b-020049bb8da4/kubernetes-dynamic-pvc-e3738adb-c569-4ed0-9773-27d2a2760f42.vmdk\" using backingDiskId :\"d3797a25-7dee-46e8-99ef-2e7f3cce5194\""}
{"level":"info","time":"2021-04-01T18:29:24.657880695Z","caller":"migration/migration.go:450","msg":"Registering volume: \"[vsanDatastore] 4bb55260-de8a-c071-913b-020049bb8da4/kubernetes-dynamic-pvc-75a78f87-0816-4e2a-b02d-0a91ceeb76a4.vmdk\" using backingDiskId :\"aea7244e-9a12-4616-99f1-7f3b29d2d3f9\""}
{"level":"info","time":"2021-04-01T18:29:29.521940801Z","caller":"migration/migration.go:450","msg":"Registering volume: \"[vsanDatastore] 4bb55260-de8a-c071-913b-020049bb8da4/kubernetes-dynamic-pvc-f2bbccd9-898c-4338-94dc-913f5d2e4a37.vmdk\" using backingDiskId :\"ead503b7-5915-4a3e-bcd8-b06d3b13ad25\""}
{"level":"info","time":"2021-04-01T18:29:35.336197569Z","caller":"migration/migration.go:450","msg":"Registering volume: \"[vsanDatastore] 4bb55260-de8a-c071-913b-020049bb8da4/kubernetes-dynamic-pvc-f35aa7b1-adef-4701-9123-7b2d720a7597.vmdk\" using backingDiskId :\"43289c14-c157-49e8-a9b8-13508176e730\""}
{"level":"info","time":"2021-04-01T18:29:41.973424865Z","caller":"migration/migration.go:450","msg":"Registering volume: \"[vsanDatastore] 4bb55260-de8a-c071-913b-020049bb8da4/kubernetes-dynamic-pvc-bbf4b95c-9966-445f-b5aa-41636c2d2967.vmdk\" using backingDiskId :\"5de132d8-9cad-41b0-b87b-d41e47094022\""}
{"level":"info","time":"2021-04-01T18:29:48.47702323Z","caller":"migration/migration.go:450","msg":"Registering volume: \"[vsanDatastore] 4bb55260-de8a-c071-913b-020049bb8da4/kubernetes-dynamic-pvc-be57b788-f24a-438e-9f39-d3e0a8db6c86.vmdk\" using backingDiskId :\"90874f52-ed66-4bd4-8ccd-12f9ba61d7e3\""}
{"level":"info","time":"2021-04-01T18:29:54.439301674Z","caller":"migration/migration.go:450","msg":"Registering volume: \"[vsanDatastore] 4bb55260-de8a-c071-913b-020049bb8da4/kubernetes-dynamic-pvc-cd9f926c-a8b3-43a9-8221-98f4103640dc.vmdk\" using backingDiskId :\"28fbb7dc-67c6-47a6-955d-38b2e806419f\""}
{"level":"info","time":"2021-04-01T18:29:59.95411649Z","caller":"migration/migration.go:450","msg":"Registering volume: \"[vsanDatastore] 4bb55260-de8a-c071-913b-020049bb8da4/kubernetes-dynamic-pvc-5d9c7764-70c6-4bd6-82da-6148ca944708.vmdk\" using backingDiskId :\"51851516-3c53-42d7-b25c-dd7f8848107f\""}
```